jetty-server/src/main/java/org/eclipse/jetty/server/PlainOrSslConnectionFactory.java

```diff
@@ -0,0 +1,192 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2019 Mort Bay Consulting Pty. Ltd.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.server;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.io.AbstractConnection;
+import org.eclipse.jetty.io.Connection;
+import org.eclipse.jetty.io.EndPoint;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * <p>A ConnectionFactory whose connections detect whether the first bytes are
+ * TLS bytes and upgrades to either a TLS connection or to a plain connection.</p>
+ */
+public class PlainOrSslConnectionFactory extends AbstractConnectionFactory
```

Comment:
I don't love the name... unless "Plain" really means plain text rather than just other.
How about "OptionalSslConnectionFactory" and the other protocol should be taken from the SslConnectionFactory rather always configured, so that you can't end up with different protocols if ssl or not?

Reply:
Plain means "non-TLS".

We cannot take the `plainProtocol` from the `SslConnectionFactory` because e.g. in case of ALPN that would be `alpn` while we want `plainProtocol` to be `http/1.1`.
Furthermore, `plainProtocol=null` is used to "abort" or "report" a plain text connection to a TLS port, so we really need this additional parameter.

I like the name "PlainOrSsl" because it also work well with the name string we give to the protocol name: `plain|ssl`, which is short and it's always printed when the connector starts.

Reply:
I don't like `plain`.

Reply:
What do you like then?
`SslOrOtherConnectionFactory`?